### ISSUE TITLE
Mongoose missed error signature in schema.post('save', fn(error, doc, next) {}) middleware

### ISSUE THREAD

**Issue by valeriy-devpronet** [DefinitelyTyped/DefinitelyTyped#12943]
- [ x ] I tried using the latest `mongoose/mongoose.d.ts` `"@types/mongoose": "^4.6.1"` file in this repo and had problems.
- [ x ] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ x ] I want to talk about `mongoose/mongoose.d.ts`.
  - The authors of that type definition are cc/ @simonxca

Hi, there is the problem with mongoose middleware definitions:
According to the docs http://mongoosejs.com/docs/middleware.html function signature for the hook post save is may looks like this (error: MongoError, doc: Document, next: (err?: NativeError) => void)
```javascript
schema.post('save', function(error, doc, next) {
  if (error.name === 'MongoError' && error.code === 11000) {
    next(new Error('There was a duplicate key error'));
  } else {
    next(error);
  }
});

```

But typings has this

```javascript
    /**
     * Defines a post hook for the document
     * Post hooks fire on the event emitted from document instances of Models compiled
     *   from this schema.
     * @param method name of the method to hook
     * @param fn callback
     */
    post<T extends Document>(method: string, fn: (doc: T, next: (err?: NativeError) => void,
      ...otherArgs: any[]) => void): this;
    post<T extends Document>(method: string, fn: (doc: T) => void, ...args: any[]): this;
```

in other words, `error` signature was omitted

**Comment by simonxca** [comment]
@valeriy-devpronet Good catch, I think I'll have to change to:
```typescript
post(method: string, fn: Function): this;
```

In these two cases:
```typescript
schema.post('save', function (doc, next) { ... });
schema.post('save', function (error, doc, next) { ... });
```
`post()` has the same number of arguments (two) so Typescript will always match the first version and enforce the 2nd parameter to be `function (doc, next) { ... }`. This will throw an error if you try to use `function (error, doc, next) { ... }`.

**Comment by valeriy-devpronet** [comment]
I'm not sure whether it's correct or not, but for me it works.
```javascript

export interface SchemaHooks extends Schema
{
  /**
     * Defines a post hook for the document
     * Post hooks fire on the event emitted from document instances of Models compiled
     *   from this schema.
     * @param method name of the method to hook
     * @param fn callback
     */
    post<T extends Document>(method: string, fn: (error, doc: T, next: (err?: NativeError) => void,
      ...otherArgs: any[]) => void): this;
    post<T extends Document>(method: string, fn: (doc: T, next: (err?: NativeError) => void,
      ...otherArgs: any[]) => void): this;
    post<T extends Document>(method: string, fn: (doc: T) => void, ...args: any[]): this;
}

```

**Comment by simonxca** [comment]
@valeriy-devpronet How did you use/import your `SchemaHooks` interface in your code? In other words, how did you override `schema.post()` using that interface?

I gave it a try by directly replacing the post method in the Schema class in `mongoose.d.ts`:
```typescript
class Schema extends events.EventEmitter {
  ...
  post<T extends Document>(method: string, fn: (error: any, doc: T, next: (err?: NativeError) => void,
    ...otherArgs: any[]) => void): this;
  post<T extends Document>(method: string, fn: (doc: T, next: (err?: NativeError) => void,
    ...otherArgs: any[]) => void): this;
  post<T extends Document>(method: string, fn: (doc: T) => void, ...args: any[]): this;
  ...
}
```

In the test code I tried these three variations:
```typescript
schema.post('post', function (doc) {
  // doc: any
}).post('post', function (doc, next) {
  // doc: any, next: mongoose.Document
}).post('post', function (error, doc, next) {
  // error: any, doc: mongoose.Document, next: (err?: NativeError) => void
});
```

The 2nd variation in particular will cause problems. Can you try the 2nd variation in your code and see if the types show up correctly?

**Comment by valeriy-devpronet** [comment]
Hi, I'm not sure that I can see this problem with my code, but I'm newbie in TypeScript so just share my code. 

```javascript

import { Schema, Model, Document, NativeError } from 'mongoose';

export interface ISchemaHooks extends Schema
{
  /**
     * Defines a post hook for the document
     * Post hooks fire on the event emitted from document instances of Models compiled
     *   from this schema.
     * @param method name of the method to hook
     * @param fn callback
     */
    post<T extends Document>(method: string, fn: (error: Error, doc: T, next: (err?: NativeError) => void,
      ...otherArgs: any[]) => void): this;
    post<T extends Document>(method: string, fn: (doc: T, next: (err?: NativeError) => void,
      ...otherArgs: any[]) => void): this;
    post<T extends Document>(method: string, fn: (doc: T) => void, ...args: any[]): this;
}

const  SchemaHooks: ISchemaHooks = new Schema({title: {type: String}});

SchemaHooks
.post('save', function (error, doc, next) {})
.post('save', function (doc, next){})
.post('save', function (doc) {})
;

SchemaHooks
.post('save', function (doc, next){})
.post('save', function (error, doc, next) {})
.post('save', function (doc) {})
;

SchemaHooks
.post('save', function (doc, next){})
.post('save', function(doc) {})
.post('save', function (error, doc, next) {})
;

SchemaHooks
.post('save', function (doc) {})
.post('save', function (doc, next){})
.post('save', function (error, doc, next) {})
;
```
![selection_016](https://cloud.githubusercontent.com/assets/11663800/20703852/fff9a344-b625-11e6-8267-11e0ae3d977e.png)

works without problem. Maybe I do not have linter which is properly but if I set wrong interface than I see that error

![selection_017](https://cloud.githubusercontent.com/assets/11663800/20703895/297e967a-b626-11e6-98a6-f15d66983c93.png)

**Comment by simonxca** [comment]
@valeriy-devpronet in your IDE, are you able to hover over the arguments `doc` and `next` in:
```typescript
...
.post('save', function (doc, next){})
...
```
And see what types they have? It should be `doc: mongoose.Document` and `next: (err?: NativeError) => void`, rather than `doc: Error` and `next: mongoose.Document`.

Also, try to compile with this code:
```typescript
...
.post('save', function (doc, next){
  doc.model;
  doc.save();
})
...
```
Since the first argument is a `mongoose.Document` it should have those properties and not throw an error.

**Comment by valeriy-devpronet** [comment]
@simonxca 
Without type 
```javascript
...
.post('save', function (doc, next){})
...
```
 shows up `(parameter) doc: any` .

If I set type mongoose.Document 
```javascript
...
.post('save', function (doc: Document, next){})
...
```
than it shows up `(parameter) doc: mongoose.Document` 

I'm using `VS Code Version 1.7.2`

For me this is fine and works as expected because I can manipulate with types.

I think that this is tied to the fact that an `error` in this typing does not have type Error

```javascript
...
post<T extends Document>(method: string, fn: (error, doc: T, next: (err?: NativeError) => void,
      ...otherArgs: any[]) => void): this;
...
```

But if I set up type `Error`

```javascript

...
post<T extends Document>(method: string, fn: (error: Error, doc: T, next: (err?: NativeError) => void,
      ...otherArgs: any[]) => void): this;
...

```

Than I get warning in the IDE about types mismatch.

I've tried to make `error` type as an optional but TypeScript does not allow to me to do this.

**Comment by simonxca** [comment]
@valeriy-devpronet Thanks for your input. Could you give these updated definitions a try:

master branch: [mongoose/mongoose.d.ts](https://raw.githubusercontent.com/simonxca/DefinitelyTyped/patch-mongoose/mongoose/mongoose.d.ts)
types-2.0 branch: [@types/mongoose/index.d.ts](https://raw.githubusercontent.com/simonxca/DefinitelyTyped/patch-mongoose-types-2.0/mongoose/index.d.ts)

For 3 arguments:
```typescript
.post('save', function (error, doc, next) { ... })
```
Should have the right types.

For 2 or less, you can declare the types:
```typescript
.post('save', function (doc: mongoose.Document, next: Function) { ... })
.post('save', function (doc: mongoose.Document) { ... })
```

**Comment by valeriy-devpronet** [comment]
Sorry for such newbie question but how can I do this using my `npm`?

**Comment by simonxca** [comment]
@valeriy-devpronet I believe you can copy paste the types-2.0 definitions above into `node_modules/@types/mongoose/index.d.ts`

Are you talking about:
`.post('save', function (error: MongoError...` or
`.post('save', function (error, doc, next: (err?: MongoError...`?

To be honest I'm not sure what the exact error type is. I wasn't able to find specific docs on it so I left it as `NativeError`, which is the same thing as `global.Error` meaning it should support `MongoError` and any other types of errors.

**Comment by valeriy-devpronet** [comment]
@simonxca 
Hi, looks good but you can see the error above.
![selection_20161201_b63c5c8](https://cloud.githubusercontent.com/assets/11663800/20789297/750c9fd2-b7bc-11e6-8a5e-da45fb630250.png)
MongoError has `name` and the `code` propereties unfortunately I can not find a complete list of these codes here is what I found

```javascript
        BAD_VALUE                 => 2,
        UNKNOWN_ERROR             => 8,
        NAMESPACE_NOT_FOUND       => 26,
        EXCEEDED_TIME_LIMIT       => 50,
        COMMAND_NOT_FOUND         => 59,
        WRITE_CONCERN_ERROR       => 64,
        NOT_MASTER                => 10107,
        DUPLICATE_KEY             => 11000,
        DUPLICATE_KEY_UPDATE      => 11001, # legacy before 2.6
        DUPLICATE_KEY_CAPPED      => 12582, # legacy before 2.6
        UNRECOGNIZED_COMMAND      => 13390, # mongos error before 2.4
        NOT_MASTER_NO_SLAVE_OK    => 13435,
        NOT_MASTER_OR_SECONDARY   => 13436,
        CANT_OPEN_DB_IN_READ_LOCK => 15927,
```
[Source](http://search.cpan.org/~mongodb/MongoDB-v1.6.0/lib/MongoDB/Error.pm)
So in other words, MongoError interface in scope of mongoose has to have `code` property, because we can get what caused an error.

**Comment by simonxca** [comment]
@valeriy-devpronet I see, the `MongoError` interface comes from `mongodb.d.ts`. Found here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/types-2.0/mongodb/index.d.ts#L31

You can ask the people who maintain it if they want to add it, but I don't think they will. The reason is because if you go to the public API docs here:
http://mongodb.github.io/node-mongodb-native/2.1/api/MongoError.html

There is no `MongoError.code` property and DefinitelyTyped will only support what is available on the public API.
http://definitelytyped.org/guides/contributing.html#private-apis

You will have to use:
```typescript
error[code] === 11000
```

**Comment by valeriy-devpronet** [comment]
@simonxca  Ok, than I think the problem is solved. If you are agree with me than you can close the ticket)

Thanks
